Xamarin.Forms.Core/Shell/Shell.cs

```diff
@@ -343,40 +351,159 @@ void IShellController.UpdateCurrentState(ShellNavigationSource source)
 
 		public static Shell Current => Application.Current?.MainPage as Shell;
 
-		Uri GetAbsoluteUri(Uri relativeUri)
+
+		[DebuggerDisplay("Full = {FullUri}, Short = {ShortUri}")]
+		class RequestDefinition
 		{
-			if (CurrentItem == null)
-				throw new InvalidOperationException("Relative path is used after selecting Current item.");
+			public RequestDefinition(Uri fullUri, Uri shortUri, ShellItem item, ShellSection section, ShellContent content)
+			{
+				FullUri = fullUri;
+				ShortUri = shortUri;
+				Item = item;
+				Section = section;
+				Content = content;
+			}
 
-			var parseUri = Regex.Match(relativeUri.OriginalString, @"(?<u>.+?)(\?(?<q>.+?))?(#(?<f>.+))?$").Groups;
-			var url = parseUri["u"].Value;
-			var query = parseUri["q"].Value;
-			var fragment = parseUri["f"].Value;
+			public RequestDefinition(string fullUri, string shortUri, ShellItem item, ShellSection section, ShellContent content) :
+				this(new Uri(fullUri, UriKind.Absolute), new Uri(shortUri, UriKind.Absolute), item, section, content)
+			{
+			}
 
-			Element item = CurrentItem;
-			var list = new List<string>();
-			while (item != null && !(item is IApplicationController))
+			public Uri FullUri { get; }
+			public Uri ShortUri { get; }
+			public ShellItem Item { get; }
+			public ShellSection Section { get; }
+			public ShellContent Content { get; }
+		}
+
+		[DebuggerDisplay("RequestDefinition = {Request}, StackRequest = {StackRequest}")]
+		class NavigationRequest
+		{
+			public enum WhatToDoWithTheStack
 			{
-				var route = Routing.GetRoute(item)?.Trim('/');
-				if (string.IsNullOrEmpty(route))
-					break;
-				list.Insert(0, route);
-				item = item.Parent;
+				ReplaceIt,
+				PushToIt
+			}
+
+			public NavigationRequest(RequestDefinition definition, WhatToDoWithTheStack stackRequest, string query, string fragment)
+			{
+				StackRequest = stackRequest;
+				Query = query;
+				Fragment = fragment;
+				Request = definition;
+			}
+
+			public WhatToDoWithTheStack StackRequest { get; }
+			public string Query { get; }
+			public string Fragment { get; }
+			public RequestDefinition Request { get; }
+		}
+
+		List<RequestDefinition> BuildAllTheRoutes()
+		{
+			List<RequestDefinition> routes = new List<RequestDefinition>();
+			// todo make better maybe
```

Comment:
If this function need only for search `RequestDefinition` by FullUri or ShortUri then it maybe better only return this RequestDefinition when finding matches with uri.
```C#
RequestDefinition GetRequestRoute(string FullUri, string ShortUri)
{
... 
if(longUri == FullUri || shortUri == ShortUri)
return new RequestDefinition(longUri, shortUri, item, section, content);
}
```